test/helpers/test_provisioner.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require_relative './kubeclient_helper'
+
+class TestProvisioner
+  extend KubeclientHelper
+
+  class << self
+    def prepare_cluster
+      WebMock.allow_net_connect!
+      $stderr.print "Preparing test cluster... "
+      prepare_pv("pv0001")
+      prepare_pv("pv0002")
+      deploy_metric_server
+      $stderr.puts "Done."
+      WebMock.disable_net_connect!
+    end
+
+    def claim_namespace(test_name)
+      test_name = test_name.gsub(/[^-a-z0-9]/, '-').slice(0, 36) # namespace name length must be <= 63 chars
+      ns = "k8sdeploy-#{test_name}-#{SecureRandom.hex(8)}"
+      create_namespace(ns)
+      ns
+    rescue KubeException => e
+      retry if e.to_s.include?("already exists")
+      raise
+    end
+
+    def delete_namespace(namespace)
+      kubeclient.delete_namespace(namespace) if namespace.present?
+    rescue KubeException => e
+      raise unless e.to_s.include?("not found")
+    end
+
+    private
+
+    def create_namespace(namespace)
+      ns = Kubeclient::Resource.new(kind: 'Namespace')
+      ns.metadata = { name: namespace }
+      kubeclient.create_namespace(ns)
+    end
+
+    def prepare_pv(name)
+      existing_pvs = kubeclient.get_persistent_volumes(label_selector: "name=#{name}")
+      return if existing_pvs.present?
+
+      pv = Kubeclient::Resource.new(kind: 'PersistentVolume')
+      pv.metadata = {
+        name: name,
+        labels: { name: name }
+      }
+      pv.spec = {
+        accessModes: %w(ReadWriteOnce),
+        capacity: { storage: "150Mi" },
+        hostPath: { path: "/data/#{name}" },
+        persistentVolumeReclaimPolicy: "Recycle"
+      }
+      kubeclient.create_persistent_volume(pv)
+    end
+
+    def deploy_metric_server
+      # Set-up the metric server that the HPA needs https://github.com/kubernetes-incubator/metrics-server
+      logger = KubernetesDeploy::FormattedLogger.build("default", KubeclientHelper::TEST_CONTEXT, $stderr)
+      kubectl = KubernetesDeploy::Kubectl.new(namespace: "kube-system", context: KubeclientHelper::TEST_CONTEXT,
+        logger: logger, log_failure_by_default: true, default_timeout: '5s')
+
+      Dir.glob("test/setup/metrics-server/*.{yml,yaml}*").map do |resource|
+        found = kubectl.run("get", "-f", resource, log_failure: false).last.success?
```

Comment:
how come you sometimes use `kubectl` and sometimes `kubeclient`? How do you decide when to use which?

Reply:
Some things we use are features of `kubectl` specifically rather than Kubernetes (most notably the `apply` verb, for now anyway), which forces the decision. Originally everything used kubectl because of that. However, kubeclient is decent now, and it is often cleaner to use it for the things it does support. So I slightly favour Kubeclient unless using it would make the code a lot more complicated, which is the case here (we'd need to load/parse the file instead of just passing its name as an arg).

That said, this was a copy-paste and we have an open issue to improve this code by paying more attention to the resources it creates. That will likely require us to parse the files anyway, so perhaps we'll switch it to Kubeclient after all.